Update LSP15 to include validityTimestamps (#223)
lukso-network/LIPs: LSPs/LSP-15-TransactionRelayServiceAPI.md

## Abstract

The [LSP-6-KeyManager](./LSP-6-KeyManager.md) proposes an [`executeRelayCall()`](./LSP-6-KeyManager.md#executerelaycall) function. It allows anybody to execute `_calldata` payload on a set ERC725 X or Y smart contract, given they have a signed message from a valid executor. This opens the way to Transaction Relay Services which send transactions on behalf of a user to cover their gas costs.

This document describes the API for a Transaction Relay Service.

## Motivation

Standardizing the Transaction Relay Service API enables applications to be compatible with all Transaction Relay Services which may be built, and avoids a situation where specific applications are only compatible with specific Transaction Relay Services. This is essential for an open marketplace of Transaction Relay Services where a user can select the service which best fits their needs.

## Specification

### API

#### POST `/execute`

Executes a signed transaction on behalf of a Universal Profile using `executeRelayCall`.

> - `address` - The Address of the Universal Profile which is executing the transaction.

Reviewer: ```suggestion
- `address` - The address of the Universal Profile which is executing the transaction.
```

> - `transaction` - An object containing the transaction parameters which will executed with `executeRelayCall`.

Reviewer 2: "be executed"

>   - `abi` - The encoded transaction data which will be passed as the payload parameter to the `executeRelayCall` function.

Reviewer: ```suggestion
  - `abi` - The abi-encoded transaction data (_e.g: a function call on the Universal Profile smart contract_) which will be passed as the payload parameter to the `executeRelayCall` function.
```

  - `signature` - The signed message according to LSP6 specification.

>   - `nonce` - The nonce of the KeyManager fetched by calling `getNonce(address address, uint256 channel)` on the LSP6 KeyManager contract.

Reviewer: Parameter type has changed for `getNonce(address,uint128)`

```suggestion
  - `nonce` - The nonce of the KeyManager fetched by calling `getNonce(address address, uint128 channelId)` on the LSP6 KeyManager contract.
```

>   - `validityTimestamps` (optional) - Two concatenated uint128 timestamps which indicate a time duration for which the transaction will be considered valid. If no validityTimestamps parameter is passed the relayer should assume that validityTimestamps is `0` and the transaction will be valid indefinitely until it is executed.

Reviewer: ```suggestion
  - `validityTimestamps` (optional) - Two concatenated `uint128` timestamps which indicate a time duration for which the transaction will be considered valid. If no validityTimestamps parameter is passed the relayer should assume that validityTimestamps is `0` and the transaction will be valid indefinitely until it is executed.
```

>   - `validityTimestamps` (optional) - Two concatenated `uint128` timestamps which indicate a time duration for which the transaction will be considered valid. If no validityTimestamps parameter is passed the relayer should assume that validityTimestamps is `0` and the transaction will be valid indefinitely until it is executed.

Reviewer 2: nice

  "transaction": {
    "abi": "0x7f23690c5ef83ad9559033e6e941db7d7c495acdce616347d28e90c7ce47cbfcfcad3bc5000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000000596f357c6aa5a21984a83b7eef4cb0720ac1fcf5a45e9d84c653d97b71bbe89b7a728c386a697066733a2f2f516d624b43744b4d7573376741524470617744687a32506a4e36616f64346b69794e436851726d3451437858454b00000000000000",
    "signature": "0x43c958b1729586749169599d7e776f18afc6223c7da21107161477d291d497973b4fc50a724b1b2ab98f3f8cf1d5cdbbbdf3512e4fbfbdc39732229a15beb14a1b",
    "nonce": 1, // KeyManager nonce

>     "validityTimestamps": "0x0000000000000000000000006420f3f000000000000000000000000065ec82d0"

Reviewer: I also wonder if it would not be easier to mention that this param can be provided as a `BigNumber` or `number`, since we are mentioning:

> If no validityTimestamps parameter is passed the relayer should assume that validityTimestamps is `0` and the transaction will be valid indefinitely until it is executed.

Author: as this is used in a REST context I think its safer to assume it will just be a string
